cluster-support-bot.py

```diff
@@ -47,53 +40,62 @@ def dashboard_uri(cluster):
-
-    user_arg_line, body = (text.strip()+'\n').split('\n', 1)
-    user_args = user_arg_line.split()[1:]  # split and drop the '<@{bot-id}>' prefix
     try:
-        args = parser.parse_args(user_args)
-    except HelpRequest as error:
-        handler = handle_help(slack_client=slack_client, event=event_data, subparser=error.args[0]['parser'])
-    except ValueError as error:
-        handler = handle_parse_args_error(slack_client=slack_client, event=event_data, error=error)
-    else:
-        handler = args.func
-        if not handler:
-            logger.info('no handler found for {!r}'.format(user_args))
+        data = payload.get('data', [])
+        if not data:
             return
-        response = handler(slack_client=slack_client, event=event_data, args=args, body=body)
-        if not response:
+        if data.get('subtype') is not None:
+            return  # https://api.slack.com/events/message#message_subtypes
+        text = data.get('text', [])
+        if not text:
+            return
+        bot_mention = "@" + bot_user
+        if bot_mention not in text:
             return
-        if response.get('ok'):
-            logger.debug(response)
-        else:
-            logger.error(response)
 
+        logger.debug('handle_message: {}'.format(payload))
 
-def handle_parse_args_error(slack_client, event, error):
-    channel = event['event']['channel']
-    thread = event['event'].get('thread_ts', event['event']['ts'])
+        timestamp = float(data.get('ts', 0))
+        if timestamp in recent_events:  # high-resolution timestamps should have few false-negatives
+            logger.info('ignoring duplicate message: {}'.format(message))
+            return
+
+        recent_events.add(timestamp)  # add after check without a lock should be a small race window
+        cutoff = time.time() - 60*60  # keep events for an hour
+        recent_events = {timestamp for timestamp in recent_events if timestamp > cutoff}
+
+        user_arg_line, body = (text.strip()+'\n').split('\n', 1)
+        user_args = user_arg_line.split()[1:]  # split and drop the '<@{bot-id}>' prefix
+        try:
+            args = parser.parse_args(user_args)
+        except HelpRequest as error:
+            handler = handle_help(payload=payload, subparser=error.args[0]['parser'])
+        except ValueError as error:
+            handler = handle_parse_args_error(payload=payload, error=error)
+        else:
+            handler = args.func
+            if not handler:
+                logger.info('no handler found for {!r}'.format(user_args))
+                return
+            response = handler(payload=payload, args=args, body=body)
+            if not response:
+                return
+            if response.get('ok'):
+                logger.debug(response)
+            else:
+                logger.error(response)
+    except Exception as e:
+        logger.debug('uncaught Exception in handle_message: {}'.format(e))
```

Comment:
I think uncaught exceptions should be `logger.error`.  And to make the scoping clearer, maybe:

```python
@slack.RTMClient.run_on(event='message')
def handle_message(**payload):
    try:
        _handle_message(payload=payload)
    except Exception as e:
        logger.debug('uncaught Exception in handle_message: {}'.format(e))


def _handle_message(payload):
    global recent_events
    ...
```

Reply:
done